### PR TITLE
[Merged by Bors] - Optimise payload attributes calculation and add SSE

### LINKED ISSUE #3700: New payload SSE event
## Description

It might be worth adding this before it's merged into the spec, I think it's a commonly desired feature https://github.com/ethereum/beacon-APIs/issues/244 - but if we implement it prior to the spec is merged we should probably make it an additional topic as opposed to modifying the head event

### LINKED ISSUE #3896: Cache withdrawals prior to block production
## Description

Presently we do a state advance for every block produced with withdrawals, here:

https://github.com/sigp/lighthouse/blob/51088725fbbe2d3cdb3f21eb968561ec5ffc1553/beacon_node/beacon_chain/src/beacon_chain.rs#L4703-L4718

This happens during `fcU` when constructing the payload attributes, so it's kind of already off the hot path. However, we _recalculate_ the withdrawals for every call.

To fix this I think there are a few options:

1. Try to use the advanced state from the snapshot cache. Downsides: state advance happens at 9s which is usually _after_ the first payload preparation call at 8s, and the state advance only gives a single slot of advancement (doesn't work for multiple skips).
2. Compute the withdrawals the same way we do currently the first time the function is called (at 8s) and cache them in the `execution_layer`. Re-wire the function to attempt to load them from the `execution_layer` on subsequent calls (e.g. for the fork choice runs at 11.5s and 0s). The withdrawals are already stored in the `proposers` map on the execution layer.

I think I prefer the 2nd solution at the moment.

We may also be able to benefit from a cache for block processing, although with the sweep length capped at 16K the iteration shouldn't take very long (by my estimate <2ms, even for `tree-states`). For block processing we'd probably want the cache as a `(Hash256, Vec<Withdrawal>)` on the `BeaconState`.

Block processing:

https://github.com/sigp/lighthouse/blob/51088725fbbe2d3cdb3f21eb968561ec5ffc1553/consensus/state_processing/src/per_block_processing.rs#L527-L528

### LINKED ISSUE #3998: Capella proposer preparation clones a beacon state
## Description

In `BeaconChain::prepare_beacon_proposer`, we're always cloning the state in order to compute the list of withdrawals: 

https://github.com/sigp/lighthouse/blob/9a41f65b892e4c679d353ddc00e87bd93b521026/beacon_node/beacon_chain/src/beacon_chain.rs#L4764-L4773

It's nice to avoid state clones since they're generally slow (10s of millis) and use a bunch of RAM (100s of MBs). I think there's at least one case were we can avoid the the state clone; when producing a block in the same epoch as its parent. Hopefully there's some more cases too.

We could also look at some sort of caching here to avoid repeat calls to the same method.

I think this behavior is OK for Sepolia, however I think we should optimise it before mainnet.

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Closes #3896
Closes #3998
Closes #3700

## Proposed Changes

- Optimise the calculation of withdrawals for payload attributes by avoiding state clones, avoiding unnecessary state advances and reading from the snapshot cache if possible.
- Use the execution layer's payload attributes cache to avoid re-calculating payload attributes. I actually implemented a new LRU cache just for withdrawals but it had the exact same key and most of the same data as the existing payload attributes cache, so I deleted it.
- Add a new SSE event that fires when payloadAttributes are calculated. This is useful for block builders, a la https://github.com/ethereum/beacon-APIs/issues/244.
- Add a new CLI flag `--always-prepare-payload` which forces payload attributes to be sent with every fcU regardless of connected proposers. This is intended for use by builders/relays.

For maximum effect, the flags I've been using to run Lighthouse in "payload builder mode" are:

```
--always-prepare-payload \
--prepare-payload-lookahead 12000 \
--suggested-fee-recipient 0x0000000000000000000000000000000000000000
```

The fee recipient is required so Lighthouse has something to pack in the payload attributes (it can be ignored by the builder). The lookahead causes fcU to be sent at the start of every slot rather than at 8s. As usual, fcU will also be sent after each change of head block. I think this combination is sufficient for builders to build on all viable heads. Often there will be two fcU (and two payload attributes) sent for the same slot: one sent at the start of the slot with the head from `n - 1` as the parent, and one sent after the block arrives with `n` as the parent.

Example usage of the new event stream:

```bash
curl -N "http://localhost:5052/eth/v1/events?topics=payload_attributes"
```

## Additional Info

- [x] Tests added by updating the proposer re-org tests. This has the benefit of testing the proposer re-org code paths with withdrawals too, confirming that the new changes don't interact poorly.
- [ ] Benchmarking with `blockdreamer` on devnet-7 showed promising results but I'm yet to do a comparison to `unstable`.

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 1]
```diff
@@ -4808,7 +4893,9 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
 
         // If we are close enough to the proposal slot, send an fcU, which will have payload
         // attributes filled in by the execution layer cache we just primed.
-        if till_prepare_slot <= self.config.prepare_payload_lookahead {
+        if self.config.always_prepare_payload
```
Comment (paulhauner) [inline beacon_node/beacon_chain/src/beacon_chain.rs:4896]: I'm wondering why we want to start updating the EE earlier when we have `always_prepare_payload == true`? I don't think it's necessarily bad, just wondering.

**Review by paulhauner — APPROVED** [review]
Looks great, this is a big improvement on what we had for a few reasons.

As discussed in another channel, I think we should get this merged before the Goerli upgrade. IMO it introduces a small amount of risk to the Goerli upgrade but it will provide valuable useful testing for mainnet.

I just had one minor question.

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs:4896]: It helps in the case of a skipped slot or 1-slot re-org where the builder wants to potentially start building a payload on the parent block as soon as the slot begins. I.e. build a slot `n + 1` block on the block from slot `n - 1` at the start of slot `n`.

Comment (michaelsproul) [reply beacon_node/beacon_chain/src/beacon_chain.rs:4896]: I guess it's a bit redundant though as builders are probably setting the payload lookahead to 12s anyway

Comment (paulhauner) [reply beacon_node/beacon_chain/src/beacon_chain.rs:4896]: Seems reasonable to me! It's an opt-in flag so I can't see any change in behavior here being a serious issue!

File: beacon_node/beacon_chain/src/events.rs [hunk 2]
```diff
@@ -43,35 +45,63 @@ impl<T: EthSpec> ServerSentEventHandler<T> {
...
-            EventKind::ChainReorg(reorg) => self.chain_reorg_tx.send(EventKind::ChainReorg(reorg))
-                .map(|count| trace!(self.log, "Registering server-sent chain reorg event"; "receiver_count" => count)),
-            EventKind::ContributionAndProof(contribution_and_proof) => self.contribution_tx.send(EventKind::ContributionAndProof(contribution_and_proof))
-                .map(|count| trace!(self.log, "Registering server-sent contribution and proof event"; "receiver_count" => count)),
-            EventKind::LateHead(late_head) => self.late_head.send(EventKind::LateHead(late_head))
-                .map(|count| trace!(self.log, "Registering server-sent late head event"; "receiver_count" => count)),
-            EventKind::BlockReward(block_reward) => self.block_reward_tx.send(EventKind::BlockReward(block_reward))
-                .map(|count| trace!(self.log, "Registering server-sent contribution and proof event"; "receiver_count" => count)),
+                .send(kind)
+                .map(|count| log_count(count, "attestation")),
```
Comment (centromere) [inline beacon_node/beacon_chain/src/events.rs:68]: Are the arguments for `log_count` reversed?

```rust
let log_count = |name, count| {
```
vs.
```rust
.map(|count| log_count(count, "attestation")),
```

Comment (michaelsproul) [reply beacon_node/beacon_chain/src/events.rs:68]: Yeah, good catch! Would you like to open a PR? Or I'm happy to do it

Comment (centromere) [reply beacon_node/beacon_chain/src/events.rs:68]: https://github.com/sigp/lighthouse/pull/4060

**Pull request closed without merging**